Create TESTING.md (#68)
usnistgov/metaschema-xslt: TESTING.md

# Testing metaschema-xslt

An important secondary goal of this project is that its processes should be *validable*, which implies transparent and auditable testing processes.

Testing in this context is very broad, including application-level testing of functionalities; testing of behaviors against documented specifications (both in this repository and elsewhere); and lower-level functional tests.

Any directory (folder) may also have its own `TESTING.md` file with information and instructions, in addition to its README file and any other documentation.

> Like each of these, this document constitutes work in progress, and contributors are invited offer edits, corrections, clarifications and updates.

Reviewer: Missing word: should say "invited to offer"

## Linting and syntax checks

Currently we are not running formal syntax checks over XSLT or XProc, or imposing any kind of rigorous syntax requirements. Developers should use good judgment and write the most legible code possible, commenting where useful.

If code legibility and consistency become an issue, we can consider more stringent rules along with automation to support them.

## *Test Everything*

TBD - a TEST EVERYTHING subroutine.

## Global functional testing

`src/testing` includes resources for global-level testing. This folder or its contents should not to be moved or edited without fully testing *all* test runtimes, as resources inside this directory are sometimes dependencies.

Do not commit anything to this folder that you do not wish to stay there indefinitely; instead, copy into a sibling (temporary) directory that can be deleted freely.

## Application component-level (functional) unit testing

`src/**/testing` include (functional) testing for utilities supported in a given folder.

> When developing applications, feel free to add and modify any `testing` folder or its contents within the scope of work.

Reviewer: Would the team consider a stronger policy recommendation about including component-level unit testing along with changes in the components? One possibility is something like, "If you touch a particular unit of code that doesn't have tests, try to test it in the same PR as your change. If you touch a particular unit of code that has tests, update or augment them to test the XSLT change you are making." The first rule can feel like it's slowing you down, but you often get the time back when you (a) spend less time dealing with bugs; or (b) can make subsequent code changes more quickly with the confidence that they don't break existing behavior.

Author: Absolutely, please check the latest revision just committed. Thanks @galtm for the words, ideas and encouragement.

Reviewer: I looked at 769e51629a34ed93a34aed5b18f1377e93c4cdad and I really like how you augmented my suggestion with additional context and motivation. Nice job!

> Unit tests are expected to run successfully when committed.

Reviewer: Do you want to add that if any XSpec tests get ahead of the implementation, they should be marked as `pending` until the implementation catches up?

Author: Yup, thanks.

## Test applications

`examples` (tbd) includes top-level independent metaschema examples made for testing and demonstration.

This location is available for lightweight and 'toy' applications, useful for demonstrations and learning. Fully built-out applications of Metaschema can also call this repository in as a submodule (like [OSCAL](https://github.com/usnistgov/oscal))

## Testing under CI/CD

> Also tbd

Reviewer: FYI, I have successfully used https://github.com/nkutsche/xspec-maven-plugin to run XSpec tests in GitHub Actions. For example, see this [pom.xml file](https://github.com/galtm/xslt-accumulator-tools/blob/db1c6b2ad0932968c2f2f4195888567bbd83c1ab/pom.xml#L68).

Author: I pulled out those links for later reference, thanks, awesome.

# Testing metaschema-xslt

> An important secondary goal of this project is that its processes should be *validable*, which implies transparent and auditable testing processes.

Reviewer 2: Minor, NB: Should this be *validatable*?
```suggestion
An important secondary goal of this project is that its processes should be *validatable*, which implies transparent and auditable testing processes.
```

Author: Hm, lemme look that up....

I looked it up and (whew) it turns out that 'validable' is indeed [correctly formed](https://english.stackexchange.com/questions/37964/word-for-something-that-can-be-validated), as I thought, from the Latin root -- but admittedly rare.

Yet as so often, a problem that isn't real reveals one that is. The sentence can be reworded:

```suggestion
An important secondary goal of this project is that its processes should be reliable, which implies transparent and auditable testing processes.
```

nb I have avoided the word "demonstrability".

Testing in this context is very broad, including application-level testing of functionalities; testing of behaviors against documented specifications (both in this repository and elsewhere); and lower-level functional tests.

> Any directory (folder) may also have its own `TESTING.md` file with information and instructions, in addition to its README file and any other documentation.

Reviewer 2: NB: Stylistic consistency
```suggestion
Any directory (folder) may also have its own `TESTING.md` file with information and instructions, in addition to its `README.md` file and any other documentation.
```

## *Test Everything*

TBD - a TEST EVERYTHING subroutine.

## Testing technologies

[XSpec](https://github.com/xspec/xspec/) is the preferred testing harness for XSLT in this initiative. XSpec test suites can be either self-contained, or can reference testing artifacts maintained externally. The repository contains XSpec examples demonstrating a range of usage patterns that can be applied.

Script-driven testing should rely on the same dependencies as the runtimes they test, as documented.

## Global functional testing

`src/testing` includes resources for global-level testing. This folder or its contents should not to be moved or edited without fully testing *all* test runtimes, as resources inside this directory are sometimes dependencies.

Do not commit anything to this folder that you do not wish to stay there indefinitely; instead, copy into a sibling (temporary) directory that can be deleted freely.

## Application component-level (functional) unit testing

`src/**/testing` includes (functional) testing for utilities supported in a given folder.

When developing applications, feel free to add and modify any `testing` folder or its contents within the scope of work.

> Unit tests are expected to run successfully when committed - both completing, and passing all applicable tests. Keep in mind that most testing frameworks support marking tests as not applicable (in XSpec, [flag a scenario or `expect` as `pending`](https://github.com/xspec/xspec/wiki/Focusing-Your-Efforts#marking-scenario-or-expectation-as-pending)), so it is possible to write tests ahead of an implementation and still pass.

Reviewer 2: Comment: This is a great feature, we should start making use of pending tests for our existing XSpec content in the OSCAL repository.

### Test-driven development

While this project began as an experimental proof of concept, it now aims for higher levels of assurance and confidence than are necessary or appropriate for applications intended only to produce findings regarding feasibility and levels of effort. Accordingly, our development approach has shifted from rapid prototyping to a more explicit and traceable process of design, specification and implementation.

> If you touch a particular unit of code that doesn't have tests, write tests for it in the same PR as your change. If you touch a particular unit of code that has tests, update or augment them to test the change you are making. In general, push the tests ahead of the code, not the other way around, aligning the tests with specifications first.

Reviewer 3: Minor, non-blocking: we might want to identify which specifications, with a link or otherwise.

```suggestion
If you touch a particular unit of code that doesn't have tests, write tests for it in the same PR as your change. If you touch a particular unit of code that has tests, update or augment them to test the change you are making. In general, push the tests ahead of the code, not the other way around, aligning the tests with [the Metaschema specification(s)](https://pages.nist.gov/metaschema/specification/) first.
```
